docs/Contributing/Blueprints/Fleet-Installers.md

```diff
@@ -0,0 +1,431 @@
+To that end, we believe the best option is Option B, implementing the functionality as a separate service.
+
+## Packager Service
+
+The "Fleet Packager" service will implement all the package generation logic. Think of it like offering the `fleetctl package` command functionality but as a REST service.
+
+```mermaid
+graph LR;
+    A[User-Agent/<br>Browser]-- Fleet API -->B[Fleet<br>UI/Server]
+    A-- Packager<br>API -->C
+    subgraph FleetDM Hosted Infrastructure
+        direction LR
+        subgraph <br>pkg.fleetctl.com
+            direction TB;
+            C[Packager<br>Service]
+            packages[(Generated<br>packages)]
+        end
+        direction TB;
+        D[tuf.fleetctl.com]    
+        C-- Fetch<br>Targets -->D;
+    end
+    E[Apple's <br> Notary <br> Server]
+    C-- Notary<br>API ---->E
+```
+
+Configurations:
+- The `Fleet UI/Server` will allow configuring the "Fleet Packager" URL (default value being FleetDM Hosted Packager service, something like `pkg.fleetctl.com`).
+- The `Packager Service` will allow configuring an alternative TUF server URL and TUF roots via an environmental variable (default value being FleetDM Hosted TUF, `tuf.fleetctl.com`). 
+
+Both of these configurations will allow users to deploy their own `Packager Service`.
+
+### Storage
+
+The generated packages should be stored on encrypted disk and will expire (will be deleted) after a configurable amount of time (default 30m).
+There are two reasons we want to expire generated packages soon:
+1. To not store user credentials for too long (URL and enroll secret).
+2. To free up space.
+
+We can use "Storage Optimized" instances (see https://aws.amazon.com/ec2/instance-types/).
+
+PS: As a possible future optimization, we could use "Memory Optimized" instances and store packages in RAM instead of using hard disk.
+
+#### Back of the Envelope
+
+- `.pkg`s use ~70MB of storage.
+- `.msi`s use ~20MB of storage.
+- `.deb`s and `.rpm`s use ~75MB of storage.
+
+Assuming the worst case of ~75 MB for each package:
+If we have a ~30TB hard disk, it would allow storing ~400_000 packages simultaneously.
+
+### Network & Credentials
+
+The service will require network access to (URLs provided via config):
+
+- TUF server.
+- Apple's Notary Server (for generating `.pkg`).
+
+The packaging service will need the following credentials (provided via config):
+
+- Apple credentials:
+  - Codesign identity.
+  - Username and Password for notarization.
+- TUF server update roots (default will be the hardcoded one for FleetDM's hosted TUF server, tuf.fleetctl.com).
+
+### Packager REST API
+
+For the MVF (Minimum Viable Feature) we'll need three APIs: one for creation/submission, one for checking status and another one for the actual download of the package. 
+All the APIs must be rate-limited to prevent abuse of the system.
+
+#### 1. Package creation
+
+`POST /create`
```

Comment:
can I issue this request to retry a build for a package that failed?

Reply:
Good question. Yes, we should. Too many moving pieces here (failures will happen, e.g. Apple servers are down, or slow, or notarization failed, or TUF server down etc.)

We could add an extra request argument to `/create` to allow the UI to re-generating a package upon failure (the service would only allow this if a previous generation failed).

So the UI flow (in the case of package generation failure) would be:
1. Do POST /create
2. Receive package_id
3. Do GET /status with package_id
4. Receive status that it failed to generate.
(At this point the UI would display the error and allow to regenerate)
5. Do POST /create (with regenerate=true)
6. ...

Am all ears here, this is not set in stone yet.

Reply:
Also, `package_id`s that failed to generate should stay in the cache longer.

Reply:
this sounds good to me 👍